### ISSUE TITLE
Extra-pass to clean up <p>

### ISSUE THREAD

**Issue by tlienart** [tlienart/Franklin.jl#677]
As noted by several users, when doing slightly sophisticated things with new commands etc, spurious `</p>` or `<p>` may be inserted. This can be (somewhat) alleviated by ensuring whitespace between "blocks" but is still not enough in some cases.

One reason for this is that it can be hard to figure out whether the content to be inserted when resolving (e.g.) a command should close a `<p>` or not. In particular, block resolution is not aware of the current context (i.e. whether a `<p>` is currently open or not). 

The situation could be improved a bit though by tagging these situations with something like `<?p>` and figuring out in a final pass whether there's currently a `<p>` open in which case that `<?p>` becomes `</p>` or there isn't in which case that `<?p>` disappears. 

### Test cases

These are examples that have been reported by users

```
\newcommand{\mwe}[1]{
~~~
<div>
~~~
`````julia
!#1
`````
~~~
</div>
~~~
}
```

(from https://gist.github.com/fredrikekre/5e123b209f96bfdc640e1866eb091414) 

(**todo** add more test cases here, @ all if you have encountered such troubles with `<p>` please kindly add them here so that they can become test cases)

**Comment by stephen-huan** [comment]
I agree that an extra post-processing step might be more robust and the correct thing to do, but in my experience for (*most*) things that one wants to do, it's possible to achieve it without introducing extraneous `<p>`'s.

For example, I can replicate the provided example
````latex
\newcommand{\mwe}[1]{
~~~
<div>
~~~
```julia
!#1
```
~~~
</div>
~~~
}

\mwe{
f(x) = 1
}
````
which renders incorrectly as
```html
<div>
</p>
<pre><code class="julia hljs">f(x) = <span class="hljs-number">1</span></code></pre>
<p>
</div>
```
(note the incorrect `</p>` at the beginning and the extra `<p>` at the end).

However, a key observation is that the div blocks are actually quite robust.
At least for this example, using a div block instead of the html as follows
````latex
\newcommand{\mwe}[1]{
@@class
```julia
!#1
```
@@
}
````
results in the correct render of
```html
<div class="class"><pre><code class="julia hljs">f(x) = <span class="hljs-number">1</span></code></pre></div>
```

Of course, this only works for `<div>`s, preventing the use of arbitrary html tags and constructions.
However, one can just recycle the existing div block [code](https://github.com/tlienart/Franklin.jl/blob/8b6f0c7d4d0f397e2ec3f5260e3a99e6bf13e939/src/converter/markdown/blocks.jl#L35-L37) into a wrapper environment called `wrap`.
```julia
"""
    env_wrap(com, _)

Wrap the contents in a html tag without spurious <p> and </p>'s.

See: [convert_md](@ref), [reprocess](@ref).
"""
function env_wrap(com, _)
    tag_data = stent(com.braces[1])
    tag, data... = split(tag_data, " ")
    data = (length(data) > 0) ? " $(join(data, ' '))" : ""
    content = stent(com)
    lxdefs = collect(values(Franklin.GLOBAL_LXDEFS))
    parsed = Franklin.reprocess(content, lxdefs;
                                nostripp=true) |> Franklin.simplify_ps
    return "~~~<$tag$data>$parsed</$tag>~~~"
end
```

Using this newly defined environment like so 
````latex
\newcommand{\mwe}[1]{
\begin{wrap}{div}
```julia
!#1
```
\end{wrap}
}
````
also results in the correct render
```html
<div><pre><code class="julia hljs">f(x) = <span class="hljs-number">1</span></code></pre></div>
```

There is another less general way to accomplish the same thing for simpler insides,
for example, consider this example with the markdown quote `> ` with the same issue. 
```latex
\newcommand{\mwe}[1]{
~~~
<div>
~~~
> !#1
~~~
</div>
~~~
}

\mwe{
testing
}
```
which results again in the incorrect render
```html
<div>
</p>
<blockquote>
<p>testing</p>
</blockquote>
<p>
</div>
```

However, because the inside is relatively simple, the markdown
construction can be directly replaced by the `<blockquote>` tag.
```latex
\newcommand{\mwe}[1]{
~~~
<div>
~~~
~~~
<blockquote>!#1</blockquote>
~~~
~~~
</div>
~~~
}
```
resulting in the correct render
```html
<div>
 
<blockquote>testing</blockquote>
 
</div>
```
this is less general because often the inside is too complicated
to make this direct substitution as we saw with the julia code fences.

Of course the `wrap` environment works on this example without
needing to change the markdown blockquote syntax `> ` 
```latex
\newcommand{\mwe}[1]{
\begin{wrap}{div}
> !#1
\end{wrap}
}
```
resulting in 
```html
<div><blockquote>
<p>testing</p>
</blockquote></div>
```

Having whitespace (specifically newlines) between blocks is also important.
The following example using `wrap` twice in a row without a newline
```latex
\newcommand{\mwe}[1]{
\begin{wrap}{div}
> !#1
\end{wrap}
}

\mwe{
test1
}
\mwe{
test2
}
```
results in the incorrect render
```html
<p><div><blockquote>
<p>test1</p>
</blockquote></div> <div><blockquote>
<p>test2</p>
</blockquote></div></p>
```
(a `<div>` cannot be placed in a `<p>` like this)

However, simply adding a newline between the two commands
```latex
\mwe{
test1
}

\mwe{
test2
}
```
results in the correct render
```html
<div><blockquote>
<p>test1</p>
</blockquote></div>
<div><blockquote>
<p>test2</p>
</blockquote></div>
```

Using these tricks `wrap` is actually quite general, for example,
it supports the relatively sophisticated `proof` environment  
```latex
\newenvironment{proof}{
  \begin{wrap}{div class="proof"}
  _Proof_.
}{
  \begin{wrap}{span class="qedsymbol"}$ \square $\end{wrap}
  \end{wrap}
}

\begin{proof}
this is a proof
\end{proof}
```
which renders as
```html
<div class="proof"><em>Proof</em>. this is a proof <span class="qedsymbol"><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi mathvariant="normal">□</mi></mrow><annotation encoding="application/x-tex"> \square </annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.675em;"></span><span class="mord amsrm">□</span></span></span></span></span></div>
```

Note that both markdown `_italic_` syntax is used as well as `$ \KaTeX $` syntax without issues. Also a `span` is used to wrap the qedsymbol, which would not be possible with the `@@ ... @@` div block syntax. Lastly, the `\end{wrap}` which ends the original `div` is delayed to the end of the environment rather than ending immediately.
